### PR TITLE
AF-2252 : File download does not work with & character

### PR DESCRIPTION
**Pull request by Rikkola**
https://issues.jboss.org/browse/AF-2252

https://github.com/kiegroup/appformer/pull/804
https://github.com/kiegroup/kie-wb-common/pull/2967
https://github.com/kiegroup/drools-wb/pull/1262

### CODE REVIEW

**Comment by domhanak** [comment]
@Rikkola Who should review this? :)

**Comment by Rikkola** [comment]
@domhanak Paulo took a quick look already, but I also need a QA to review :)

**Comment by Rikkola** [comment]
@paulovmr I think this fix might be the best. The `encodeQueryString `method worked if there were no spaces, but with spaces the download broke again. This was because spaces were encoded, but & character was not. So now I first `decode `and then `encodeQueryString`.

**Comment by domhanak** [comment]
Getting the war

Roses are red,
violets are blue,
Jenkins execute full downstream build!

**Comment by paulovmr** [comment]
Jenkins execute full downstream build

**Comment by Rikkola** [comment]
> [FATAL] Non-resolvable parent POM for org.jboss.errai:errai-parent:4.7.0-SNAPSHOT

Jenkins execute full downstream build

File: uberfire-extensions/uberfire-commons-editor/uberfire-commons-editor-client/src/main/java/org/uberfire/ext/editor/commons/client/menu/DownloadMenuItemBuilder.java [hunk 1]
```diff
@@ -74,6 +75,7 @@ MenuItem makeMenuItem(final String caption,
     }
 
     private String getFileDownloadURL(final Supplier<Path> pathSupplier) {
-        return GWT.getModuleBaseURL() + "defaulteditor/download?path=" + pathSupplier.get().toURI();
+
+        return GWT.getModuleBaseURL() + "defaulteditor/download?path=" + URIUtil.encodeQueryString(URIUtil.decode(pathSupplier.get().toURI()));
```
Comment (sthundat) [inline uberfire-extensions/uberfire-commons-editor/uberfire-commons-editor-client/src/main/java/org/uberfire/ext/editor/commons/client/menu/DownloadMenuItemBuilder.java]: "defaulteditor/download?path=" can we make this literal  as a constant

Comment (Rikkola) [reply uberfire-extensions/uberfire-commons-editor/uberfire-commons-editor-client/src/main/java/org/uberfire/ext/editor/commons/client/menu/DownloadMenuItemBuilder.java]: Reasonable suggestion and done.

**Comment by sthundat** [comment]
Not able to download decision table  with & in name for example 'demo&decisiontable' .There are 2 download buttons are available for decision table.One is downloading the file as expected and another download option is throwing exception .

``` 'org.uberfire.java.nio.IOException
	at org.guvnor.common.services.backend.file.upload.AbstractFileServlet.processAttachmentDownload(AbstractFileServlet.java:309)
	at org.guvnor.common.services.backend.file.upload.AbstractFileServlet.doGet(AbstractFileServlet.java:273)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:686)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:791).
```

Please find attachment for more information


DecisionTableIssue

**Comment by Rikkola** [comment]
Uploading with the character also broken. Taking a look.

**Comment by Rikkola** [comment]
I still need to clean up, please hold the rereviews :)

File: uberfire-server/src/main/java/org/uberfire/server/FileUploadServlet.java [hunk 2]
```diff
@@ -115,11 +120,4 @@ private void finalizeResponse(HttpServletResponse response,
         writeResponse(response,
                       RESPONSE_OK);
     }
-
-    private String getExtension(final String originalFileName) {
-        if (originalFileName.contains(".")) {
-            return "." + originalFileName.substring(originalFileName.lastIndexOf('.') + 1);
-        }
-        return "";
-    }
```
Comment (Rikkola) [inline uberfire-server/src/main/java/org/uberfire/server/FileUploadServlet.java]: This was dead code.

Comment (Rikkola) [inline uberfire-server/src/main/java/org/uberfire/server/FileUploadServlet.java]: This was dead code.

File: uberfire-server/src/main/java/org/uberfire/server/FileUploadServlet.java [hunk 3]
```diff
@@ -51,29 +49,10 @@ protected void doPost(HttpServletRequest request,
...
-
-                final FileItem fileItem = getFileItem(request);
-
-                finalizeResponse(response,
-                                 fileItem,
-                                 uri);
-            }
+            finalizeResponse(response,
+                             getFileItem(request),
+                             getTargetLocation(request));
```
Comment (Rikkola) [inline uberfire-server/src/main/java/org/uberfire/server/FileUploadServlet.java]: All this logic moved to `UploadUriProvider.java`

**Comment by Rikkola** [comment]
@sthundat @paulovmr @domhanak Had to expand the fix for upload and download, both had problems with the & character. It looks like some resources call the Upload/Dowload servlets with encoded content and for this reason I'm for safety reason encoding before I encode again.

Also I noticed that reuploading a file with the default upload editor does not work since the resource does not exits. Opened a separate ticket for that here:
https://issues.jboss.org/browse/DROOLS-4675

**Comment by savithajean** [comment]
@Rikkola Please take a look on the failed tests.

**Comment by sthundat** [comment]
@Rikkola Can you please take a look on the failed tests

**Comment by Rikkola** [comment]
Fixing. I'm encoding a bit too much when a file is updated.

**Comment by Rikkola** [comment]
Updated, can't run the appformer repo tests locally in Windows so need to use jenkins to see if all works.

**Comment by domhanak** [comment]
jenkins restest this

**Comment by sthundat** [comment]
Jenkins execute full downstream build

**Comment by sutaakar** [comment]
Jenkins execute full downstream build

**Comment by Rikkola** [comment]
> Unexpected error while executing the Downstream repositories builder for PR builds! PR 2967 for repo GitHubRepository{owner='kiegroup', name='kie-wb-common'} is not automatically mergeable. Please fix the conflicts first!
> java.lang.RuntimeException: PR 2967 for repo GitHubRepository{owner='kiegroup', name='kie-wb-common'} is not automatically mergeable. Please fix the conflicts first!

**Comment by Rikkola** [comment]
Jenkins execute full downstream build

**Comment by Rikkola** [comment]
Jenkins execute full downstream build

**Comment by Rikkola** [comment]
> org.drools.workbench.screens.dtablexls.backend.server.conversion.GuidedDecisionTableGeneratorListenerTest.testAttributes
> org.drools.workbench.screens.guided.dtable.client.widget.analysis.DecisionTableAnalyzerTest.testInsertNonExistingColumn
> org.drools.workbench.screens.guided.dtable.client.widget.analysis.DecisionTableAnalyzerTest.testInsertActionColumn
> org.drools.workbench.screens.guided.dtable.client.widget.analysis.DecisionTableAnalyzerTest.testInsertBRLVariableColumn
> org.drools.workbench.screens.guided.dtable.client.widget.analysis.DecisionTableAnalyzerTest.testInsertBRLConditionColumn
> org.drools.workbench.screens.guided.dtable.client.widget.table.model.synchronizers.impl.ActionRetractFactColumnSynchronizerTest.testMoveColumnTo_MoveLeft
> org.drools.workbench.screens.guided.dtable.client.widget.table.model.synchronizers.impl.ActionRetractFactColumnSynchronizerTest.testMoveColumnTo_MoveRight
> org.drools.workbench.screens.guided.dtable.client.widget.table.model.synchronizers.impl.ActionRetractFactColumnSynchronizerTest.testMoveColumnTo_OutOfBounds
> org.drools.workbench.screens.guided.dtable.client.widget.table.model.synchronizers.impl.ConditionColumnSynchronizerTest.testMoveColumnsTo_MoveLeft_MidPoint
> org.drools.workbench.screens.guided.dtable.client.widget.table.model.synchronizers.impl.ConditionColumnSynchronizerTest.testUpdate2

Inspecting what broke these.

**Comment by Rikkola** [comment]
Jenkins execute full downstream build

**Comment by mareknovotny** [comment]
failed `org.kie.bc.DeploymentIT.org.kie.bc.DeploymentIT`
Trying again as that test are flaky.

Jenkins execute full downstream build

**Comment by Rikkola** [comment]
Ok so there is indeed some issue with this when downloading a file with &, but no spaces.

**Comment by Rikkola** [comment]
Wasn't about spaces/no spaces. There was yet another location where we encode download URLs.
Jenkins execute full downstream build

**Comment by Rikkola** [comment]
Jenkins execute full downstream build

**Comment by sthundat** [comment]
Verified and issue is resolved now

**Comment by Rikkola** [comment]
@sthundat Thanks for the reveiews.

**Comment by Rikkola** [comment]
@kiegroup/gatekeepers Can I get a merge please.

**Pull request merged**